07-folds-monoids/JoinList.hs

```diff
@@ -0,0 +1,244 @@
+(!!?) :: [a] -> Int -> Maybe a [] !!? _ = Nothing
+!!? i | i < 0 = Nothing (x:xs) !!? 0 =Justx (x:xs) !!? i = xs !!? (i-1)
+which returns Just the ith element in a list (starting at zero) if such an element exists, or Nothing otherwise. We also consider an updated function for converting join-lists into lists, just like jlbToList but ignoring the monoidal annotations:
+jlToList :: JoinList m a -> [a]
+jlToList Empty = []
+jlToList (Single _ a) = [a]
+jlToList (Append _ l1 l2) = jlToList l1 ++ jlToList l2
+We can now specify the desired behavior of indexJ. For any index i and join-list jl, it should be the case that
+   (indexJ i jl) == (jlToList jl !!? i)
+That is, calling indexJ on a join-list is the same as first convert- ing the join-list to a list and then indexing into the list. The point, of course, is that indexJ can be more efficient (O(log n) versus O(n), assuming a balanced join-list), because it gets to use the size annotations to throw away whole parts of the tree at once, whereas the list indexing operation has to walk over every element.
+-}
+indexJ :: (Sized b, Monoid b) => Int -> JoinList b a -> Maybe a
+indexJ n (Single _ a)
+    | n == 0    = Just a
+    | otherwise = Nothing
+indexJ n (Append m jl1 jl2)    
+    | n < jl1Size = indexJ n             jl1
+    | otherwise   = indexJ (n - jl1Size) jl2
+      where
+        jl1Size :: Int
+        jl1Size = getSize . size . tag $ jl1
+indexJ _ _ = Nothing
+
+
+
+
+{-
+Implement the function
+dropJ :: (Sized b, Monoid b) =>
+Int -> JoinList b a -> JoinList b a
+The dropJ function drops the first n elements from a JoinList. This is analogous to the standard drop function on lists. Formally, dropJ should behave in such a way that
+jlToList (dropJ n jl) == drop n (jlToList jl)
+-}
+dropJ :: (Sized b, Monoid b) => Int -> JoinList b a -> JoinList b a
+dropJ n (Single m a)
+    | n >= 1    = Empty
+    | otherwise = Single m a
+dropJ n (Append _ jl1 jl2)
+    | n < jl1Size = dropJ n             jl1 +++ jl2
+    | otherwise   = dropJ (n - jl1Size) jl2
+      where
+        jl1Size :: Int
+        jl1Size = getSize . size . tag $ jl1
+dropJ _ _ = Empty
+
+
+
+
+{-
+Finally, implement the function takeJ :: (Sized b, Monoid b) =>
+Int -> JoinList b a -> JoinList b a
+The takeJ function returns the first n elements of a JoinList, dropping all other elements. Again, this function works similarly to the standard library take function; that is, it should be the case that
+jlToList (takeJ n jl) == take n (jlToList jl).
+Ensure that your function definitions use the size function from the Sized type class to make smart decisions about how to descend into the JoinList tree.
+-}
+takeJ :: (Sized b, Monoid b) => Int -> JoinList b a -> JoinList b a
+takeJ n (Single m a)
+    | n <= 0    = Empty
+    | otherwise = Single m a
+takeJ n (Append _ jl1 jl2)
+    | n < jl1Size = takeJ n jl1
+    | otherwise   = jl1 +++ takeJ (n - jl1Size) jl2
```

Comment:
I think this solution has the same issue that I pointed out in the above comment. See if you can improve this solution!